File: library/src/main/java/me/proxer/library/api/ConferenceInfoAdapter.java
package me.proxer.library.api;

import com.squareup.moshi.FromJson;
import com.squareup.moshi.Json;
import me.proxer.library.entitiy.messenger.ConferenceInfo;
import me.proxer.library.entitiy.messenger.ConferenceParticipant;

import java.util.Date;
import java.util.List;

/**
 * @author Ruben Gees
 */
class ConferenceInfoAdapter {

    @FromJson
    ConferenceInfo fromJson(final IntermediateConferenceInfoContainer json) {
        return new ConferenceInfo(json.info.topic, json.info.participantAmount, json.info.firstMessageTime,
                json.info.lastMessageTime, json.info.leaderId, json.participants);
    }

    private static class IntermediateConferenceInfoContainer {

        @Json(name = "conference")
        IntermediateConferenceInfo info;

        @Json(name = "users")
        List<ConferenceParticipant> participants;

        static class IntermediateConferenceInfo {

            @Json(name = "topic")
            String topic;

            @Json(name = "count")
            int participantAmount;

            @Json(name = "timestamp_start")
            Date firstMessageTime;

            @Json(name = "timestamp_end")
            Date lastMessageTime;

            @Json(name = "leader")
            String leaderId;
        }
    }
}

File: library/src/main/kotlin/me/proxer/library/internal/adapter/ConferenceInfoAdapter.kt
package me.proxer.library.internal.adapter

import com.squareup.moshi.FromJson
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import me.proxer.library.entity.messenger.ConferenceInfo
import me.proxer.library.entity.messenger.ConferenceParticipant

import java.util.Date

/**
 * @author Ruben Gees
 */
internal class ConferenceInfoAdapter {

    @FromJson
    fun fromJson(json: IntermediateConferenceInfoContainer): ConferenceInfo {
        return ConferenceInfo(
            json.info.topic, json.info.participantAmount, json.info.firstMessageTime,
            json.info.lastMessageTime, json.info.leaderId, json.participants
        )
    }

    @JsonClass(generateAdapter = true)
    internal data class IntermediateConferenceInfoContainer(
        @Json(name = "conference") internal val info: IntermediateConferenceInfo,
        @Json(name = "users") internal val participants: List<ConferenceParticipant>
    ) {

        @JsonClass(generateAdapter = true)
        internal data class IntermediateConferenceInfo(
            @Json(name = "topic") val topic: String,
            @Json(name = "count") val participantAmount: Int,
            @Json(name = "timestamp_start") val firstMessageTime: Date,
            @Json(name = "timestamp_end") val lastMessageTime: Date,
            @Json(name = "leader") val leaderId: String
        )
    }
}
